### DISCUSSION TITLE
Allowed characters in a stream address (Filename, folders)

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3861] [General]
Hello, All.

I have run into some character issues and am in the process of renaming album folders.
The song names are taken care of, but some album names are an issue.

`UPDATE`
This is regarding the directory file text for the liquidsoap running on Ubuntu.
example: `https://demo.example.com/host/Files/Band_Name/1984-Album+name/08202023185330903.mp3`
The above will give an error before of the `+` in the album_name.

Names like this are not allowed.
Error received `Error while fetching http data: 523 - Origin is unreachable`
Away_From_The_Sun_(Special_Edition) <-- parentheses
This,_Is_a_Title <-- has a comma after the first word.
And'_this_one <-- has an apostrophe
`+`  <-- plus
`:`  <-- colon
`;`  <-- semi-colon
`.` <-- Period
`=` <-- Equal
Is there a list of allowed characters that I can put in my replace function so that when these albums are uploaded to the server, they will be given the proper names?

Thanks.
Wayne

**Comment by Moonbase59** [comment]
If you’re on Linux, strictly spoken, anything except a `NUL` character goes. For practical purposes, `/` should be excluded, since it is the folder separator character. If this _doesn’t_ work, it is because a piece of software or a script is badly written.

In my test cases, I have playlists and file names like
```
Test: "a_b"? 'ÄÖÜẞäöüß 测试 テスト Тест! *%<> \hello.flac
```

and they all work. It also depends on the underlying _file system_ used, of course.

Windows/MacOS might pose more restrictions, don’t know and don’t care much.

Two rather "okay" answers on Stack Overflow:
- https://stackoverflow.com/a/31976060
- https://stackoverflow.com/a/1976050

**Reply by carrzkiss** [reply]
Hello, Moonbase.
I updated the title and information. This is all regarding the link to the media file.

**Reply by Moonbase59** [reply]
Ah, talking about `https:` URIs! That has nothing to do with Liquidsoap, it’s just a matter of URL-encoding them properly. Within Liquidsoap, you can use [url.encode](https://www.liquidsoap.info/doc-dev/reference.html#url.encode) to do that. (Your `+`, in an URL, normally signifies a ` ` (blank)).

Like, my `Test: "a_b"? 'ÄÖÜẞäöüß 测试 テスト Тест! *%<> \hello.flac` from above, if properly URL-encoded, would look like this:

```
https://demo.example.com/host/Files/Test%3A%20%22a_b%22%3F%20%27%C3%84%C3%96%C3%9C%E1%BA%9E%C3%A4%C3%B6%C3%BC%C3%9F%20%E6%B5%8B%E8%AF%95%20%E3%83%86%E3%82%B9%E3%83%88%20%D0%A2%D0%B5%D1%81%D1%82%21%20%2A%25%3C%3E%20%5Chello.flac
```

and your example like this:

```
https://demo.example.com/host/Files/Band_Name/1984-Album%2Bname/08202023185330903.mp3
```

No problem here, almost all programming languages can properly URL-encode and URL-decode. Note [RFC3986 Section 2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) though (encode _parts_ before constructing the URI).

**Reply by carrzkiss** [reply]
Thanks, Moonbase.
Now, using the url.encode for the link paths to the song.
Let's say the song is this.
>The band name has periods between every letter.
> The album name has a period between the first and last name.
> The album name has parentheses.

```
https://demo.example.com/host/Files/B.A.N.D./1984-Album.Name/08202023185330903.mp3
https://demo.example.com/host/Files/B.A.N.D./1984-AlbumName-(Part02)/08202023185330903.mp3
```

So, this can stay as-is, with no modification to the actual name.
The only thing that needs to be done is using the url.encode on the LIQ file, is this correct?

Would the url.encode be placed here?

`music = playlist(mode='normal',reload=1, reload_mode='rounds', "https://www.example.com/DJ/88_1.m3u")`

And would I need to use the url.decode here as well?
If the only thing I need it for is to retrieve the file for streaming.

**Reply by carrzkiss** [reply]
OK, I just did the URL encoding on the server side, and it worked for the names with a period in them.
But now for the parentheses 

Actual album name (Foldname)
`https://demo.example.com/host/Files/Evanescence/2004-Anywhere_But_Home_(LIVE)/01222024174431188.mp3`
So, the URL Encode really messes up by replacing the _ underscore as well.
So, I am not sure how reliable this method is going to be.
Here is the output after doing the URL Encode.
`2004-Anywhere%5FBut%5FHome%5F%28LIVE%29/01222024174431188.mp3`
And here is the error from Liquidsoap; it did not like it too much.
```
2024/04/07 22:24:48 [procol.external:3] Failed to find a file extension for 
"https://demo.example.com/host/Files/Evanescence/2004-Anywhere%5FBut%5FHome%5F%28LIVE%29/01222024174431188.mp3". 
Response has unknown mime-type: "text/html" you may want to add it to 
`settings.http.mime.extnames` and report to us if it is a common one.
2024/04/07 22:24:49 [procol.external:3] Error while fetching http data: 404 -
2024/04/07 22:24:49 [request.dynamic:3] Fetch failed: empty.
```

I tried just encoding the parentheses, but that did not work either.
`https://demo.example.com/host/Files/Evanescence/2004-Anywhere_But_Home_%28LIVE%29/01222024174431188.mp3`
This one works. Replacing the periods
`https://demo.example.com/host/Files/W%2EA%2ES%2EP/1992-The_Crimson_Idol/08222023212114312.mp3`

**Reply by Moonbase59** [reply]
As I said:

> Note [RFC3986 Section 2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) though (encode parts before constructing the URI).

You’d need to deconstruct your path (the `/` is the path separator which _mustn’t_ be URL-encoded), URL-encode the _parts_ (except the so-called _authority_ which is the _scheme_ + _domain_ + _port_ part), then join them together again.

I don’t know where these filenames you have come from, but if it is some API or your own LS function, you could have _that_ deliver already-encoded file paths.

This all looks like possibly over-complicating things at one point or another. Maybe there is a simpler solution. **What is the actual use case for this?** I mean, of course we’re all getting lazy nowadays because browsers do _exactly that_ when you type something into the address bar. But of course that _doesn’t_ mean anything lower-level will.

How did you URL-encode your example with the _wrongly_ encoded underscores? Liquidsoap seems to work correctly here, look:

```
$ liquidsoap 'print(url.encode("2004-Anywhere_But_Home_(LIVE)"));'
2004-Anywhere_But_Home_%28LIVE%29
```

The result for the full paths should be:

```
https://demo.example.com/host/Files/Evanescence/2004-Anywhere_But_Home_%28LIVE%29/01222024174431188.mp3
https://demo.example.com/host/Files/W.A.S.P/1992-The_Crimson_Idol/08222023212114312.mp3
```

Your `W.A.S.P` example is a good example for _Windows deficiencies_: On Windows, a folder name _cannot end with a period_!

**Comment by carrzkiss** [comment]
I am making up a list of allowed characters in the folder and file names for liquidsoap to interpret.
What I did to test.
I added a few files to a folder and named the folder 
this folder
And between the two words, I added each of the characters listed below.
Restrarted the station, and if it played the song listed for that character, I marked it as good.
If it did not play, then I tossed it.
I am going to continue looking for and testing different characters to resolve the issue I am having and to help others who may have the same problem.

```
~  (tilde) 
`  (accent left) 
^  (Caret) ctrl 6
…  (Ellipsis) alt 0133
°  (Degree Sign) alt 0176
—  (Em Dash) alt 0151
–  (En Dash) alt 0150
•   (bullet) alt 0149
©  (Copyright) alt 0169
®  (Trademark) alt 0174
@  (At Symbol) shift 2
† (Dagger) Alt 0134
‡ (Dagger) Alt 0135
```

**Reply by Moonbase59** [reply]
You might not succeed building a successful exclusion table for the currently defined 149,878 code points in Unicode 15.1, out of a possible ~1.1 million.

**Reply by carrzkiss** [reply]
My main goal is to get the names working.
These characters are just the ones I have found that work.
The ones I am using are: 

`Curly brackets {Live}` -replace- `parentheses (Live)`.
And for the period, I might go with `•   (bullet)`

I am not going to use the whole set, just the ones needed to complete the task.
I am listing them here and on my website as well, so others who have similar issues can find them quicker than I did.

**Comment by carrzkiss** [comment]
I changed everything over to an ID System.
Now the folder structure is like this.

stream.website.com/host/Files/001/005/08202023185330903.mp3

001 = Band ID
005 = Album ID

So now there is no issues with trying to do what I was trying to do in here.
Should have been designed this way to begin with.
